### ISSUE TITLE
torch.cat puts result on current GPU rather than GPU of inputs

### ISSUE THREAD

**Issue by jekbradbury** [pytorch/pytorch#689]
```python
>>> import torch
>>> a = torch.rand(5).cuda(1)
>>> b = torch.rand(5).cuda(1)
>>> c = torch.cat([a,b], 0)
>>> c.get_device()
0
>>> torch.cuda.set_device(1)
>>> c = torch.cat([a,b], 0)
>>> c.get_device()
1
```
This is a problem for `nn.DataParallel` and similar multi-gpu things. Paszke says it might be a bug in [the wrapper](https://github.com/pytorch/pytorch/blob/master/torch/csrc/generic/methods/Tensor.cwrap#L576), because this is what `THCPAutoGPU` is supposed to deal with.

**Comment by colesbury** [comment]
THCPAutoGPU doesn't handle tensors nested in lists; it only looks one layer deep and the tensor arguments to cat are nested in a list.

**Comment by apaszke** [comment]
Fixed in #734.
